docs/what-is-wrapped.md

```diff
@@ -0,0 +1,1007 @@
+      Avatar(Array.empty)
+    )
+}
+
+class UserService(repo: UserRepository) {
+  def get(id: UserId): UserView = {
+    val user = repo.get(id)
+    UserView(
+      user.id,
+      user.age,
+      user.email,
+      user.gender,
+      user.name,
+      user.department,
+      user.avatar
+    )
+  }
+}
+```
+
+Much better! Small addition `extends AnyVal` will tell the compiler that it doesn't need to wrap our classes and waste resource unless it has to ( read this article for more information https://docs.scala-lang.org/overviews/core/value-classes.html )<br>
+However, it's still not perfect. We must redefine `toString` everywhere and yet again we have to take care of it manually.
+Wouldn't it be nice to have some common trait that will redefine required methods automatically?<br>
+Actually we can easily add this trait and call it `Wrapped` :) Let's see how it looks.
+```scala
+// `extends Any` doesn't have hidden meaning.
+// It is here only to bypass Scala limitation on value classes, try to remove it and see what happens
+trait Wrapped[T] extends Any {
+  def value: T
+  override def toString: String = value.toString
+}
+
+final case class UserId(value: String) extends AnyVal with Wrapped[String]
+final case class Name(value: String) extends AnyVal with Wrapped[String]
+final case class Age(value: Int) extends AnyVal with Wrapped[Int]
+final case class Gender(value: String) extends AnyVal with Wrapped[String]
+final case class Password(value: String) extends AnyVal with Wrapped[String]
+final case class Email(value: String) extends AnyVal with Wrapped[String]
+final case class Department(value: String) extends AnyVal with Wrapped[String]
+final case class Avatar(value: Array[Byte]) extends AnyVal with Wrapped[Array[Byte]]
+
+final case class UserView(
+  id: UserId,
+  age: Age,
+  email: Email,
+  gender: Gender,
+  name: Name,
+  department: Department,
+  avatar: Avatar
+)
+
+final case class User(
+  id: UserId,
+  name: Name,
+  age: Age,
+  gender: Gender,
+  password: Password,
+  email: Email,
+  department: Department,
+  avatar: Avatar
+)
+
+class UserRepository {
+  // let's forget about this unsafe place. Usually it's done by DB libraries that won't mess up
+  def get(id: UserId): User =
+    User(
+      id,
+      Name(s"some_name_$id"),
+      Age(21),
+      Gender(s"non_binary_gender"),
+      Password(s"super_secret_password_$id"),
+      Email("user@email.com"),
+      Department("user_department"),
+      Avatar(Array.empty)
+    )
+}
+
+class UserService(repo: UserRepository) {
+  def get(id: UserId): UserView = {
+    val user = repo.get(id)
+    UserView(
+      user.id,
+      user.age,
+      user.email,
+      user.gender,
+      user.name,
+      user.department,
+      user.avatar
+    )
+  }
+}
+```
+I hope now you see we have added this trait to our project.
+However, you probably noticed that our `Wrapped` contains not only `toString`, but also a bunch of other methods.<br>
```

Comment:
Can't notice that actually :)

Reply:
Ooops. It should be `real Wrapped` instead of `our Wrapped`. Fixed, thanks!